[RHCLOUD-18655] feature: add tenant translation endpoints

MikelAlejoBR opened the pull request:
We were initially going to perform another migration to translate the
leftover tenants which don't have an associated "OrgId" value, but after
a stand up meeting we decided upon creating a couple of endpoints to be
able to do it on demand.

These endpoints will help at:

1. Fetching all the "translatable" tenants.
2. Initiating the "EBS account number to OrgId" process.

Of course, these endpoints go on the internal routes, so they will only
be able to be hit from inside the cluster.

## Links

[[RHCLOUD-18655]](https://issues.redhat.com/browse/RHCLOUD-18655)

MikelAlejoBR committed:
feature: add translation endpoints

We were initially going to perform another migration to translate the
leftover tenants which don't have an associated "OrgId" value, but after
a stand up meeting we decided upon creating a couple of endpoints to be
able to do it on demand.

These endpoints will help at:

1. Fetching all the "translatable" tenants.
2. Initiating the "EBS account number to OrgId" process.

Of course, these endpoints go on the internal routes, so they will only
be able to be hit from inside the cluster.

lpichler commented on dao/tenant_dao.go:
```diff
@@ -68,3 +77,141 @@ func (t *tenantDaoImpl) TenantByIdentity(id *identity.Identity) (*m.Tenant, erro
+				// Grab the updated tenant for the "translated tenant" struct.
+				var tenant m.Tenant
+				// Update the tenant with the translated "orgId".
+				dbResult := DB.Debug().
```
I did not play with transaction in gorm but should not be here `tx` instead of `DB.Debug()` ?

MikelAlejoBR replied:
Ooof you're right, yes. I wrapped everything on a transaction and then I forgo to use it :joy: Thanks!

MikelAlejoBR committed:
refactor: use the transaction object instead of the DB one

MikelAlejoBR committed:
fix: where condition should be "OR not empty"

We're interested in "external_tenant"s that are not null or not empty,
to try and translate them.

MikelAlejoBR committed:
fix: limit the number of fetched EBS account numbers each time

It was fetching all of them because I forgot to add the "Limit" clause
to the query. Now it should respect the "translate 100 tenants at a
time" constraint.

lpichler commented on dao/tenant_dao.go:
```diff
@@ -70,3 +77,141 @@ func (t *tenantDaoImpl) TenantByIdentity(id *identity.Identity) (*m.Tenant, erro
+
+				// If the update yielded 0 affected rows that means that no tenant could be found with that
+				// "external_tenant".
+				if dbResult.RowsAffected == 0 {
```
- EAN has to exist in sources DB because of condition in `untranslatedTenantsWhereCondition`.
- IF EAN exists in sources DB but doesn't exist in translation service I guess - it is caught on line 148.
- `org_id` is not populated yes in tenant DB for the EAN

what has to happen to have this condition true ?

but probably it doesn't hurt to have it here.

MikelAlejoBR replied:
Since we are using a transaction that code should be "dead code", you're right. Do I leave it just in case? :thinking:

lpichler replied:
yes, I agree 👍

lpichler commented on dao/tenant_dao.go:
```diff
@@ -70,3 +77,141 @@ func (t *tenantDaoImpl) TenantByIdentity(id *identity.Identity) (*m.Tenant, erro
+					continue
+				}
+
+				if err != nil {
```
it looks that this `err != nil` is covered on 134 or am I missing something ?

MikelAlejoBR replied:
You're right. Actually this was meant to be `dbResult.Error`, to be able to log a potential error with the update statement.

lpichler commented on dao/tenant_dao.go:
```diff
@@ -70,3 +77,141 @@ func (t *tenantDaoImpl) TenantByIdentity(id *identity.Identity) (*m.Tenant, erro
+				// Grab the updated tenant for the "translated tenant" struct.
+				var tenant m.Tenant
+				// Update the tenant with the translated "orgId".
+				dbResult := tx.
```
Is in postgres batch update ? I think that yes - but maybe it is not needed here - I guess we don't have too much untranslated EAN.

MikelAlejoBR replied:
I thought Gorm supported batch updating, but I am facing a really weird "unsupported data" when fetching the tenants and then calling `db.Updates(&tenants)`. I've looked on the internet to actually see if you can do a batch update when you're trying to update multiple rows with different values and... things are a bit complicated.

However, aside from this, the implementation I came up with was a little bit weird:

1. I fetched 100 tenants per loop.
2. Created a `map[string]*m.Tenant` to store in memory the tenants by their external tenant.
3. Called the translation service.
4. I used the `map[string]` to update each of the tenants with their corresponding `org_id`.
5. Updated the database with the tenants.
6. Looped through the tenants to generate a proper HTTP response.

Here's the implementation: https://gist.github.com/MikelAlejoBR/54c6f3d3694556cc4344f6f8c3e03f88

I would honestly not refactor the current PR's code to the above, to be honest... plus I would have to check if there's a way to properly batch update with Gorm without hitting an error :eyes:

lpichler replied:
yes, you are right and thank you very much for trying it and for those information! 👏

lpichler commented on dao/tenant_dao.go:
```diff
@@ -70,3 +77,141 @@ func (t *tenantDaoImpl) TenantByIdentity(id *identity.Identity) (*m.Tenant, erro
+				Where(untranslatedTenantsWhereCondition).
+				Pluck("external_tenant", &ebsAccountNumbers).
+				Offset(int(i)).
+				Limit(100).
```
I guess we need to add here Order BY clause as order of records is not ensured.

MikelAlejoBR replied:
Mmmmmmmm... I don't follow. Why do we want ordered records here? :thinking: I included the clause anyways :+1:

lpichler replied:
Idea is when order of records is not ensured then it can take first 100 records(for offset=0) and then it can pick also some same records for next 100 records - 

is it probably edge case and most likely it will not cause issue - tenants from batch are processed and org_id is set and then those tenants are excluded from this query thanks to the  `untranslatedTenantsWhereCondition`
- so it pertains just records which remain after translation processing without `org_id`.

hopefully it makes sense.

Thanks for adding order by!

MikelAlejoBR replied:
Thanks an interesting point, yeah. Thanks for the explanation! :+1:

lpichler commented on dao/tenant_dao.go:
```diff
@@ -70,3 +77,141 @@ func (t *tenantDaoImpl) TenantByIdentity(id *identity.Identity) (*m.Tenant, erro
+
+func (t *tenantDaoImpl) TranslateTenants() (int64, uint64, uint64, []m.TenantTranslation, error) {
+	// The where condition to fetch only the untranslated tenants.
+	untranslatedTenantsWhereCondition := `("external_tenant" IS NOT NULL OR "external_tenant" != '') AND ("org_id" IS NULL OR "org_id" = '')`
```
is it possible to share this condition as with `GetUntranslatedTenants` ?

MikelAlejoBR replied:
Sure, I created a `const` with that string.

MikelAlejoBR committed:
refactor: check the databse error on the update statement

MikelAlejoBR committed:
refactor: add the tenant's ID to the successful INFO log

MikelAlejoBR committed:
refactor: create a constant with the common "Where" condition

MikelAlejoBR committed:
refactor: order by clause to pluck ordered EBS account numbers

MikelAlejoBR committed:
refactor: handle the database error first to vaoid shadowing it

When the query returned with an error, the "dbResult.RowsAffected" was
0, and since it was placed before the error handling, the error was
getting ignored.

lpichler reviewed (approved):
Looks great! 👍

lindgrenj6 commented on dao/tenant_dao.go:
```diff
@@ -103,3 +111,141 @@ func (t *tenantDaoImpl) TenantByIdentity(id *identity.Identity) (*m.Tenant, erro
+			}
+
+			// Batch process 100 EBS account numbers.
+			results, err := translator.EANsToOrgIDs(context.Background(), ebsAccountNumbers)
```
This will actually send all of the `ebsAccountNumbers` at once, the name of the method is misleading.

We should probably loop over this slice 100 at a time and translate them that way.

MikelAlejoBR replied:
I believe I am doing that in the code, right? Everything is wrapped on a `for i := int64(0); i < translatableTenants; i += 100` loop. Were you actually referring to fetching everything at once and then applying the loop to just the translator part?

lindgrenj6 replied:
Yep I see it now. I totally missed it on L154, sorry about that.

MikelAlejoBR replied:
No problem at all! :)
